client/network-gossip/src/state_machine.rs

```diff
@@ -308,36 +308,40 @@ impl<B: BlockT> ConsensusGossip<B> {
 				validator.validate(&mut context, &who, &message)
 			};
 
-			let validation_result = match validation {
-				ValidationResult::ProcessAndKeep(topic) => Some((topic, true)),
-				ValidationResult::ProcessAndDiscard(topic) => Some((topic, false)),
-				ValidationResult::Discard => None,
+			let (topic, keep) = match validation {
+				ValidationResult::ProcessAndKeep(topic) => (topic, true),
+				ValidationResult::ProcessAndDiscard(topic) => (topic, false),
+				ValidationResult::Discard => {
+					trace!(target:"gossip", "Discard message from peer {}", who);
+					continue;
+				},
 			};
 
-			if let Some((topic, keep)) = validation_result {
-				network.report_peer(who.clone(), rep::GOSSIP_SUCCESS);
-				if let Some(ref mut peer) = self.peers.get_mut(&who) {
-					peer.known_messages.insert(message_hash);
-
-					to_forward.push((topic, TopicNotification {
-						message: message.clone(),
-						sender: Some(who.clone())
-					}));
-
-					if keep {
-						self.register_message_hashed(
-							message_hash,
-							topic,
-							message,
-							Some(who.clone()),
-						);
-					}
-				} else {
+			network.report_peer(who.clone(), rep::GOSSIP_SUCCESS);
```

Comment:
This mimics the behaviour of the previous implementation. I doubt we want to report the gossip as success even though it might still be a peer that is not registered (see below). Am I missing something? Otherwise I will fix this in a follow up.

Reply:
+1 for putting this line after we know that the peer is connected.